MonoGame.Framework/Graphics/Image2D.cs

```diff
@@ -0,0 +1,165 @@
+        {
+            if (width <= 0)
+            {
+                throw new ArgumentOutOfRangeException("width");
+            }
+
+            if (height <= 0)
+            {
+                throw new ArgumentOutOfRangeException("height");
+            }
+
+            Width = width;
+            Height = height;
+
+            _data = new byte[width * height * 4];
+        }
+
+        /// <summary>
+        /// Constructs a <see cref="Image2D"/> of the specified size and containing specified data
+        /// </summary>
+        /// <param name="width"></param>
+        /// <param name="height"></param>
+        /// <param name="data"></param>
+        public Image2D(int width, int height, byte[] data)
+        {
+            if (width <= 0)
+            {
+                throw new ArgumentOutOfRangeException("width");
+            }
+
+            if (height <= 0)
+            {
+                throw new ArgumentOutOfRangeException("height");
+            }
+
+            if (data == null)
+            {
+                throw new ArgumentNullException("data");
+            }
+
+            var length = width * height * 4;
+            if (data.Length != length)
+            {
+                throw new ArgumentException(string.Format("Inconsistent data length: expected={0}, provided={1}", length,
+                    data.Length));
+            }
+
+            Width = width;
+            Height = height;
+            _data = data;
+        }
+
+        /// <summary>
+        /// Creates Texture2D from this image
+        /// </summary>
+        /// <param name="graphicsDevice"></param>
+        /// <returns></returns>
+        public Texture2D CreateTexture2D(GraphicsDevice graphicsDevice)
+        {
+            var texture = new Texture2D(graphicsDevice, Width, Height, false, SurfaceFormat.Color);
+            texture.SetData(_data);
+            return texture;
+        }
+
+        /// <summary>
+        /// Loads an image in PNG, JPG, BMP, GIF or TGA format from the stream
+        /// </summary>
+        /// <param name="stream"></param>
+        /// <returns></returns>
+        public unsafe static Image2D FromStream(Stream stream)
+        {
+            // Rewind stream if it is at end
+            if (stream.CanSeek && stream.Length == stream.Position)
+            {
+                stream.Seek(0, SeekOrigin.Begin);
+            }
+
+            ImageResult result;
+            if (stream.CanSeek)
+            {
+                result = ImageResult.FromStream(stream, ColorComponents.RedGreenBlueAlpha);
+            }
+            else
+            {
+                // If stream doesnt provide seek functionaly, use MemoryStream instead
+                using (var ms = new MemoryStream())
+                {
+                    stream.CopyTo(ms);
+                    ms.Seek(0, SeekOrigin.Begin);
+                    result = ImageResult.FromStream(ms, ColorComponents.RedGreenBlueAlpha);
```

Comment:
What is the idea behind making a memory stream, does StbImageSharp need to seek on the stream to read stuff?

Reply:
Yes. It is used in method stbi__skip.

Reply:
But, why don't you just modify stbi__skip to ask if the stream is seekable, if it is, move X bytes, otherwise, read X bytes.

(assuming it never goes backwards)

Reply:
It goes backwards. I just looked code again and there's a method stbi__rewind that rewinds stream to the beginning.

Reply:
Hmm, looking at the source of StbImageSharp, the only thing that actually needs backwards seeking is for figuring out the image type: https://github.com/StbSharp/StbImageSharp/blob/88b9c906dd5769d7684b6ef9227af881c00eeb9c/src/StbImage.Generated.Common.cs#L75

I'm fine with your new changes, however it would be very good if you could cash the few bytes from the header so you don't need a seeable stream.